app/services/fact_importer.rb

```diff
@@ -81,26 +82,35 @@ def add_new_facts
         end
       end
     end
-
     @counters[:added] = facts_to_create.size
-    logger.debug("Merging facts for '#{host}': added #{@counters[:added]} facts")
+    ActiveSupport::Notifications.instrument "fact_importer_added.foreman", :host_id => host.id, :host_name => host.name, :facts => facts, :count => @counters[:added], :added => facts_to_create do
+      logger.debug("Merging facts for '#{host}': added #{@counters[:added]} facts")
+    end
   end
 
   def create_fact_name(fact_names, name, fact_value)
     fact_names[name] ||= fact_name_class.create!(:name => name)
   end
 
   def update_facts
-    facts_to_update = []
-    db_facts.each { |name, fv| facts_to_update << [facts[name], fv] if fv.value != facts[name] }
-
-    @counters[:updated] = facts_to_update.size
-    return logger.debug("No facts update required for #{host}") if facts_to_update.empty?
-
-    logger.debug("Merging facts for '#{host}': updated #{@counters[:updated]} facts")
-
-    facts_to_update.each do |new_value, fv|
-      fv.update_attribute(:value, new_value)
+    updated = []
+    db_facts.find_each do |record|
+      new_value = facts[record.name]
+      if record.value != new_value
+        # skip callbacks/validations
+        record.update_column(:value, new_value)
+        updated << record
+      end
+    end
+    # set updated_at field explicitly
+    FactValue.where(:id => updated.map(&:id)).update_all(:updated_at => Time.now)
```

Comment:
after thinking about this a bit more, we're already updating each of these rows above, so why not use:
```ruby
record.update_columns(:value => new_value, :updated_at => Time.now)
```